### DISCUSSION TITLE
Tab completion conflicts [GitHub Copilot]

### DISCUSSION THREAD

**Discussion by cipherswami** [community/community#7036] [Copilot Conversations] [answered]
TAB is not completing the CoPilot suggestions,
Github Copilot is conflicting with other extensions most probably emmt I think.

My Opinion:
Github CoPilot should give us an option to assign a custom key instead of a [TAB] or should change to something like [SHIFT + TAB] instead of TAB.

**Comment by jansorg** [comment]
Which editor are you using? Thanks!

**Reply by cipherswami** [reply]
vs code

**Comment by Krzysztof-Cieslak** [comment, marked as answer]
Please check our documentation about configuring VSCode extension - https://github.com/github/copilot-docs/blob/main/docs/visualstudiocode/configuring.md

**Reply by sebhs** [reply]
that link doesn't work for me

**Reply by Paddy-Hamilton** [reply]
+1

**Reply by ssddi456** [reply]
+1

**Reply by DragonAxe01** [reply]
+1

**Reply by Sunlis** [reply]
https://docs.github.com/en/copilot/configuring-github-copilot/configuring-github-copilot-in-visual-studio-code

**Comment by subssn21** [comment]
I am having the same issue in pyCharm. Sometimes Tab Auto completes with the suggestion, Other times it autocompletes with something from another pycharm plugin (Not sure which one). Can you change it to something else, or does someone have any ideas on how to prevent the other autocomplete from taking over.

**Reply by jansorg** [reply]
Thanks! Although TAB is supposed to be handled first by GitHub Copilot, there seems to be a problem.
Could you open "Help > About", click "Copy" and then past the info here? This will help to debug which other plugin may have overridden the tab handling.
TAB should also be available for custom shortcuts, which is also under investigation.

**Reply by jansorg** [reply]
The latest update of the plugin allows to set your own shortcut. Could you try that, please? You can set up your shortcut at `Preferences > Keymap > Plugin > GitHub Copilot`.

**Reply by rexebin** [reply]
Awesome, I changed the shortcut for Github Copilot 'Apply completion to Editor' to 'alt+shift+enter', working like a dream.

**Reply by TylerMSFT** [reply]
@rexebin - thank you. Changing *Apply completion to Editor* to `alt+shift+enter` worked for me as well.

**Comment by Lshiff** [comment]
Hey I seem to be having the same problem, I don't know why but tab isn't doing anything. I'm not sure how to fix it as I have no idea what else tab is assigned to or how to change. Now it just moves forward four spaces I think

**Comment by subssn21** [comment]
PyCharm 2021.2 (Professional Edition)
Build #PY-212.4746.96, built on July 27, 2021
Licensed to Michael Robellard
Subscription is active until March 28, 2022.
Runtime version: 11.0.11+9-b1504.13 x86_64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
macOS 12.0.1
GC: ParNew, ConcurrentMarkSweep
Memory: 1979M
Cores: 16
Non-Bundled Plugins: GLSL (1.19), com.neueda4j.intellij.plugin.cypher (3.0.3), org.jetbrains.plugins.vue (212.4746.57), sourcery.pycharm-plugin (0.9.3), com.github.copilot (1.0.2)

**Comment by cipherswami** [comment]
It took a lot of time to reconfigure but was finally done.

My Suggestions for default keybinds:

  1. Accept an inline suggestion: **TAB+TAB**. `editor.action.inlineSuggest.commit`
   (as TAB by default is used by *emmt)
  
  2.Open Copilot (10 suggestions in separate pane): **ALT + PERIOD**. `github.copilot.generate`
  (as CTRL+ENTER by default is used to go to *next line) 

* important

**Reply by Lshiff** [reply]
Is this on vscode? also, do you know to disable emmt

**Reply by cipherswami** [reply]
In settings.json File add this line:
"editor.quickSuggestions": false

**Reply by cipherswami** [reply]
OR you can skip emmt suggestion by ESCP

**Reply by mufidu** [reply]
Thanks! It works for me.

**Comment by jrmajor** [comment]
I have the same issue in PhpStorm. The conflicting shortcut is “Indent Line or Selection”. When I unassign <kbd>tab</kbd> from it, Copilot autocompletion works.

**Reply by jansorg** [reply]
Thanks! This was supposed to be overridden by Copilot, but it doesn't seem to work for your setup. 
The latest update (out today) allows to setup your own shortcut instead of Tab (at `Preferences > Keymap > Plugins >GitHub Copilot`).

With the latest update, it uses a different way to override existing shortcut. Could you try again with the new version?

**Reply by cipherswami** [reply]
@jansorg 
It would be nice if VS Code has a direct option in Extension settings to change key instead of keybinding.

**Reply by jrmajor** [reply]
@jansorg I want to be able to indent my code regardless of whether Copilot has any suggestions, so I've mapped completion to <kbd>⌥</kbd> + <kbd>⇥</kbd>. Thank you!

**Reply by cipherswami** [reply]
`ALT(⌥) + ENTER` is the best option.

But I have checked the latest version, 
After resetting the keybinding to default, `editor.action.inlineSuggest.commit` is still set to TAB.

[VS Code]

**Comment by timkelty** [comment]
@jansorg I am still unable to change the "accept suggestion" key from `TAB` in PhpStorm.

I dont see `⇥` mapped anywhere in Copilot's keymap:
<img width="690" alt="CleanShot 2022-01-15 at 12 36 19@2x" src="https://user-images.githubusercontent.com/18329/149632036-57a22c95-ee95-45c2-90f0-add57bad7686.png">

> PhpStorm 2021.3.1
> GitHub Copilot plugin 1.0.2

**Reply by jansorg** [reply]
@timkelty Updating to the latest version (1.1.4.220) should fix it. Make sure to the remove the custom repository if you were using the early bird releases.

**Reply by timkelty** [reply]
🤦 forgot I had even done that. Thanks!

**Comment by bobmagicii** [comment]
yeah i must be too blind or dumb for it too - in vscode there are no bindings set and yet its consuming my tab. its really infuriating when im in a class and it auto completes just a lone `}` and im just trying to indent again for literally anything not that. all it does is cause me to get in a war where i get madder and madder doing tab backspcae tab backspace tab backspace hoping it will give up.

honestly i think the inline stuff sucks all around. i would prefer they go into the normal auto complete dropdown, or something similar (not the sidebar that opens, that is flow killing) instead of constantly messing with the DOM which does mess with all sorts of things like positioning of your cursor or where you were about to click the moment it came up with its brilliant idea of habitually creating non-valid syntax: https://pbs.twimg.com/media/FLQ2gNiXEAsFvjm?format=png&name=orig

i am often reading other code while typing code to make sure my context is right, and with inline constantly pushing crap around its just bad experience for me. turning off inline tho is just about as bad as disabling the plugin completely, because im never going to sit here and ask for the sidebar, read its ideas, audit its idea, select an idea, and then audit it again because it always guesses api that does not exist. that will never be faster than just typing it and being done. 

i need an unobtrusive way to get suggestions live, like the normal vs code suggestion dropdowns that appear near your cursor.

and since im dogpiling, an option to limit the length of a suggestion. every full function it suggests has been terrible. api that doesn't exist, formatting that is wrong for our codebase. but when it only suggests short things, like a list of function arguments, those work really really well. personally i never want the ai suggesting code that is longer than 1 line. it becomes too much a hassle to audit versus just typing a few characters and using the IDE's normal understanding of the codebase to tab complete.

**Comment by dale-c-anderson** [comment]
In PyCharm, reassigning the tab key binding works fine, but every other key combination seems a lot less convenient, unless I want to break another binding. I've chosen to just live with it for now, as this only seems to present as an issue in yml/yaml files. :thinking:

**Comment by justrustyspoons** [comment]
I have found that it's when you use a live template (in JetBrains IDEs / PHPStorm) this happens. For example in JavaScript live templates is ```log``` which resolves to  ```console.log($END$)``` when you write ```log``` and hit TAB it places your cursor between the parenthesis for the ```$END$``` variable. If copilot suggests something and you hit tab it just skips the copilot suggestion. So it does seem like a conflict with live templates TAB and copilot TAB function. I'm going to open an issue with JetBrains about it

**Comment by bigplayer-ai** [comment]
this happens as well in CLion, it indents/adds a tab before the completion of GitHub copitlot

**Comment by brandon-kohn** [comment]
For my case I use the vsvim extension. You have to be in insert mode to have it work.

**Comment by d4e666** [comment]
this issue also applies to VS2022

Microsoft Visual Studio Community 2022
Version 17.11.0
VisualStudio.17.Release/17.11.0+35208.52
Microsoft .NET Framework
Version 4.8.09037

Installed Version: Community

Visual C++ 2022   00482-90000-00000-AA463
Microsoft Visual C++ 2022

ASP.NET and Web Tools   17.11.230.51983
ASP.NET and Web Tools

Atomineer Pro Documentation   1.0
Source-code documentation-comment creation and maintenance, and other documentation-related tools.

Azure App Service Tools v3.0.0   17.11.230.51983
Azure App Service Tools v3.0.0

C# Tools   4.11.0-3.24374.1+f99bb16a395e48a05520ba7af1549b20bfdeee36
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Extensibility Message Bus   1.4.39 (main@e8108eb)
Provides common messaging-based MEF services for loosely coupled Visual Studio extension components communication and integration.

Extension Manager 2022   1.3.4
Import/export extensions as well as associate extensions with individual solutions

File Icons   2.7
Adds icons for files that are not recognized by Solution Explorer

GitHub Copilot   0.2.1627.58842
GitHub Copilot is an AI pair programmer that helps you write code faster and with less work.

JwtTools   0.2.18
A tool window to help with JWT token decoding and other things.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Mono Debugging for Visual Studio   17.11.3 (ba13144)
Support for debugging Mono processes with Visual Studio.

NuGet Package Manager   6.11.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

PlantUML Editor   1.1.133
PlantUML Editor

Razor (ASP.NET Core)   17.11.3.2436904+f07c7bb8e623ebcc1e1028cf11d101f866c77a79
Provides languages services for ASP.NET Core Razor.

SonarLint for Visual Studio   8.2.1.96841
SonarLint is an extension to your favorite IDE that provides on-the-fly feedback to developers on new bugs and quality issues injected into their code.

TypeScript Tools   17.0.30715.2002
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   4.11.0-3.24374.1+f99bb16a395e48a05520ba7af1549b20bfdeee36
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual F# Tools   17.11.0-beta.24317.3+c07c8ebad62f6629b36349381991a299598a8d74
Microsoft Visual F# Tools

Visual Studio Extension for SpecFlow   1.0
Visual Studio extension for working with SpecFlow projects and Gherkin feature files.

Visual Studio IntelliCode   2.2
AI-assisted development for Visual Studio.

VisualStudio.DeviceLog   1.0
Information about my package

VisualStudio.Mac   1.0
Mac Extension for Visual Studio

VSColorOutput64   2023.4
Color output for build and debug windows - https://mike-ward.net/vscoloroutput

Xamarin   17.11.0.93 (d17-11@885b409)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   17.11.3.10 (remotes/origin/d17-11@714146c62b)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

the issue can be reproduced by creating an eventhandler in c# using the += syntax, vs gives the implementation, but hitting tab does nothing. 

This should be standard bahvior and all the suggestiosn about changing the keymapping, I cna't find any key maping thatis linked to gihub at all.
